src/engraving/dom/edit.cpp

```diff
@@ -6471,19 +6474,43 @@ void Score::undoAddCR(ChordRest* cr, Measure* measure, const Fraction& tick)
                 // TODO solve this in read302.cpp.
                 tracks.push_back(strack);
             } else {
-                // linkedPart : linked staves within same part/instrument.
-                // linkedScore: linked staves over different scores via excerpts.
-                const bool linkedPart  = linked && (staff != ostaff) && (staff->score() == ostaff->score());
-                const bool linkedScore = linked && (staff != ostaff) && (staff->score() != ostaff->score());
-                for (track_idx_t track : muse::values(mapping, strack)) {
+                std::vector<track_idx_t> mappedTracks = muse::values(mapping, strack);
+                if (mappedTracks.empty()) {
+                    // This is a linked staff in a part and the element has been added to a linked staff in the main score
+                    track_idx_t track = staffIdx * VOICES + (strack % VOICES);
+                    track_idx_t mappedTrack = muse::value(mapping, track, muse::nidx);
+                    mappedTrack = (mappedTrack != muse::nidx) ? mappedTrack : track;
+                    if (staff->isVoiceVisible(strack % VOICES)) {
+                        mappedTracks.push_back(mappedTrack);
+                    }
+                }
+                for (track_idx_t track : mappedTracks) {
+                    // linkedPart : linked staves within same part/instrument.
+                    // linkedScore: linked staves over different scores via excerpts.
+                    const bool linkedPart  = linked && (staff != ostaff) && (staff->score() == ostaff->score());
+                    const bool linkedScore = linked && (staff != ostaff) && (staff->score() != ostaff->score());
                     if (linkedPart && !linkedScore) {
-                        tracks.push_back(staff->idx() * VOICES + muse::value(mapping, track) % VOICES);
+                        track_idx_t mappedTrack = muse::value(mapping, track, muse::nidx);
+                        if (mappedTrack == muse::nidx) {
+                            continue;
+                        }
+                        track_idx_t linkedTrack = staffIdx * VOICES + mappedTrack % VOICES;
+                        if (!staff->isVoiceVisible(strack % VOICES)) {
+                            continue;
+                        }
+                        tracks.push_back(linkedTrack);
                     } else if (!linkedPart && linkedScore) {
-                        if ((track >> 2) != staff->idx()) {
-                            track += (staff->idx() - (track >> 2)) * VOICES;
+                        // Staff in linked score
+                        if ((track >> 2) != staffIdx) {
```

Comment:
Ok, bit-shifting by 2 bits, in order to divide by 4, which is equal to VOICES, in order to get the staffIdx from a trackIdx, must be the most convoluted thing I've ever seen in this codebase 😅. Please let's take the opportunity to get rid of this and replace it with track / VOICES (or better use the track2staff function)